### PR TITLE
Instrument the app to produce metrics and traces

### PR DESCRIPTION
**Pull request by kiendang**
Part of PE-563.

- Display Prometheus metrics at `/metrics` using [`django-prometheus`](https://github.com/korfuri/django-prometheus).
- Produce traces and send them to the local Grafana Agent (specified by the env var `OTLP_GRPC_ENDPOINT`, default to `http://localhost:4317`)

### CODE REVIEW

**Comment by shuesken** [comment]
@kiendang is there any way of running this in a way that does not require additional services? We're very far from Soundcloud levels of complexity or logging needs, and we don't have an easy way of just running a Prometheus service (i.e. we'd either have to set up another machine or set up some k8s or other container env for which we could just run the image).

**Comment by kiendang** [comment]
Ah my approach is we don't run Prometheus, Jaeger or any service like that ourselves. Have been looking into the simplest way we can have observability (metrics + tracing) and my take is we register for a free Grafana account https://grafana.com/pricing/. All we have to do is to instrument our app, run a Opentelemetry collector or Grafana agent (this is just Opentelemetry collector with Grafana stuff on top). The collector is just a background service managed by systemd which collects the metrics and tracing stuff from our app and send it to Grafana via gRPC.

Grafana will manage Prometheus and Tempo (their tracing service, similar to Jaeger) for us and provide visualization dashboard as well.

So in summary what we need to set up and maintain:

- Register for a free Grafana plan.
- Instrumenting our app so it collects tracing and metrics. Metrics is alr done in this PR. Tracing is similarly simple
- Install opentelemetry collector or Grafana agent, write a yaml config file so that it collects traces and metrics from the app and send to Grafana, start the service in the same server with the app (`systemctl services start ...`).
- Log in to Grafana and view/query traces, metrics.

**Comment by kiendang** [comment]
Will finish this within the coming week.

**Comment by kiendang** [comment]
Choosing Grafana because their free plan is very usable, even allows for team of 3 to log in. Apart from the obvious visualization/dashboard, the plan provides services for all 3 areas of observability: metrics (Prometheus), tracing (Tempo) and logs (Loki, this is not included in this ticket, but if we want can very easily add in).

**Comment by kiendang** [comment]
The Otel collector/Grafana agent thing is optional even. Can send traces directly from the app to Grafana cloud. Using the collector is much more convenient and recommended though.

**Comment by shuesken** [comment]
> * Register for a free Grafana plan.
> 
> * Instrumenting our app so it collects tracing and metrics. Metrics is alr done in this PR. Tracing is similarly simple
> 
> * Install opentelemetry collector or Grafana agent, write a yaml config file so that it collects traces and metrics from the app and send to Grafana, start the service in the same server with the app (`systemctl services start ...`).
> 
> * Log in to Grafana and view/query traces, metrics.

None of this except "Register for a free plan" really means anything to me. This does sound like a bunch of things to configure and maintain, or at least to know the names of. I.e it easily doubles the number of services we're running. Is there a way of doing this in a less complex way? Either way, we'll have to carefully document it in the server setup runbook:
https://app.nuclino.com/BankGreen/bank-green/Server-setup-runbook-5430fc5e-2cfb-43c5-a361-2790972f0687

**Comment by shuesken** [comment]
Or maybe more pragmatically – can you run this in a hacky enough way for us to get a sense of whether we get any data that's at all useful to for performance improvement?

**Comment by kiendang** [comment]
Ah sorry
> - Instrumenting our app so it collects tracing and metrics.
This just means adding code to this django app so it produces metrics and traces. This is just a few lines of code. Metrics has already been done in this PR. For traces it's similarly simple.

The opentelemetry colector/Grafana agent is a background daemon to collect the metrics and traces produced by the app and sends it to grafana. All we have to do is installing a `.deb` package, configure it so that it collects metrics from our app and sends to Grafana. This is just setting the app host:port and the Grafana address and API key. Basically set a couple env vars, change a couple lines in a yaml config file and that's basically it.

> Or maybe more pragmatically – can you run this in a hacky enough way for us to get a sense of whether we get any data that's at all useful to for performance improvement?

Yup that's the plan.
